Implement FGA on SHOW COMMANDS

e-dard opened the pull request:
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [ ] CHANGELOG.md updated

This PR does the following:

   - Fix bug in `tsi1` where database names were not propagating properly. This would have been a problem for FGA.
   - Implement FGA on the `SeriesPointIterator`. This will be needed by `SHOW SERIES`, and I suspect `ifql`.
   - Add test coverage for FGA when using `SHOW TAG KEYS` and `SHOW TAG VALUES`. It looks like FGA was working on them even after the recent `SHOW` command refactor.
  - Implement FGA on `SHOW MEASUREMENTS` now that `SHOW MEASUREMENTS` is using its own command, rather than a `SelectStatement`.
  - Fix a bug in `tsi1` where negation operations on tags would not return the correct values.
  - Add significant test coverage for FGA on `SHOW MEASUREMENTS`.

benbjohnson commented on tsdb/index/inmem/inmem.go:
```diff
@@ -490,20 +493,29 @@ func (i *Index) measurementNamesByNameFilter(op influxql.Token, val string, rege
+	var authorized bool
+
+	// valEqual determins if the provided []byte is equal to the tag value
+	// to be filtered on.
```
`s/determins/determines/`

benbjohnson commented on tsdb/index/inmem/inmem.go:
```diff
@@ -490,20 +493,29 @@ func (i *Index) measurementNamesByNameFilter(op influxql.Token, val string, rege
+			return bytes.Equal([]byte(filter.Value), b)
+		}
+		return filter.Regex.Match(b)
+	}
```
Does this need to be an inline function? Can you move it out and just pass `filter` in?

benbjohnson commented on tsdb/index/inmem/inmem.go:
```diff
@@ -513,37 +525,53 @@ func (i *Index) measurementNamesByTagFilters(filter *TagFilter) [][]byte {
+		authorized = true
+		if auth != nil {
+			authorized = false // Authorization must be explicitly granted.
+		}
```
```go
authorized := auth != nil // Authorization must be explicitly granted.
```

benbjohnson commented on tsdb/store_test.go:
```diff
@@ -959,6 +960,193 @@ func TestStore_TagValues(t *testing.T) {
+		for _, tv := range values {
+			if got, exp := tv.Measurement, "cpu"; got != exp {
+				return fmt.Errorf("got measurment %q, expected %q", got, exp)
+			}
```
`s/measurment/measurement/`

joelegasse commented on tsdb/index/tsi1/file_set.go:
```diff
@@ -639,62 +643,115 @@ func (fs *FileSet) measurementNamesByNameFilter(op influxql.Token, val string, r
+			for ve := vitr.Next(); ve != nil; ve = vitr.Next() {
+				// If a match was found for the EQ or EQREGEX operator then
+				// the measurement should be included (if it's authorized).
+				if valEqual(ve.Value()) {
```
`if !valEqual() { continue }` and save the indent?

joelegasse commented on tsdb/index/tsi1/index_test.go:
```diff
@@ -394,3 +338,11 @@ func (idx *Index) CreateSeriesSliceIfNotExists(a []Series) error {
 	return nil
 }
+
+func BytesToStrings(a [][]byte) []string {
```
I see this in a couple places, would it make sense to add to the `bytesutils` package?

e-dard replied:
Sounds good. I did consider this, but was concerned others' might object to putting code only used in tests into non-test packages.

e-dard commented:
@benbjohnson @joelegasse please see a5af19fc0 where I have addressed feedback. I also added an extra optimisation on `inmem` where we only continue checking for tag matches if we're [using a regex](https://github.com/influxdata/influxdb/pull/9127/commits/a5af19fc06e6fe319f30162c14abbb88b24b3edf#diff-b27ccb4c1a2b59f94983d00fd2381c5cR551)

benbjohnson commented:
👌 Don't forget the `CHANGELOG`.
